iocore/net/SSLNetVConnection.cc

```diff
@@ -263,20 +267,11 @@ ssl_read_from_net(SSLNetVConnection *sslvc, EThread *lthread, int64_t &ret)
       Debug("ssl.error", "[SSL_NetVConnection::ssl_read_from_net] SSL_ERROR_WOULD_BLOCK(read/x509 lookup)");
       break;
     case SSL_ERROR_SYSCALL:
-      TraceIn(trace, sslvc->get_remote_addr(), sslvc->get_remote_port(), "Syscall Error: %s", strerror(errno));
       SSL_INCREMENT_DYN_STAT(ssl_error_syscall);
-      if (nread != 0) {
-        // not EOF
-        event = SSL_READ_ERROR;
-        ret   = errno;
-        Debug("ssl.error", "[SSL_NetVConnection::ssl_read_from_net] SSL_ERROR_SYSCALL, underlying IO error: %s", strerror(errno));
-        TraceIn(trace, sslvc->get_remote_addr(), sslvc->get_remote_port(), "Underlying IO error: %d", errno);
-      } else {
-        // then EOF observed, treat it as EOS
-        // Error("[SSL_NetVConnection::ssl_read_from_net] SSL_ERROR_SYSCALL, EOF observed violating SSL protocol");
-        TraceIn(trace, sslvc->get_remote_addr(), sslvc->get_remote_port(), "EOF observed violating SSL protocol");
-        event = SSL_READ_EOS;
-      }
+      event = SSL_READ_ERROR;
+      ret   = (int)(-nread);
```

Comment:
Probably passing along the return value of SSL_read is more useful than errno.  We probably need to review a bit more on how we are using this value though.  Later in one of the debugs we are calling strerror() on this value which only makes sense for OS errno values.  Actually we may just want to call SSL_get_error() at this point to fetch the proper SSL error code and pass it around.

Oh, that is what we are doing in SSLReadBuffer, and the return value of that function is the result of SSL_get_error.  So probably dragging around the nread (return value of SSL_read) or the errno is not very interesting.  I think we added the errno at various points just to help with debugging messages anyway.

Reply:
`SSL_ERROR_SSL` and `SSL_ERROR_SYSCALL` may be returned by `SSL_get_error()`.

To generate debug information for SSL_ERROR_SSL with `ERR_peek_last_error()` and `ERR_error_string_n()` while using `errno` and `strerror()` for SSL_ERROR_SYSCALL.

The `SSLNetVC::load_buffer_and_write` try to map the `SSL ERROR` into `-errno` because:
- SSLNetVC inherits from UnixNetVC,
- The `UnixNetVC::load_buffer_and_write` returns written bytes or `-errno`,
- `UnixNetVC::load_buffer_and_write` and `SSLNetVC::load_buffer_and_write` are called by `write_to_net_io()` which is defined within `UnixNetVConnection.cc`.

To implement the map, it is better to

- return `-errno` directly if the ssl error is SSL_ERROR_SYSCALL.
- otherwise build a key-value table to map SSL_ERROR_xxx into -EAGAIN or -EPIPE.